text/0000-async.md

```diff
@@ -0,0 +1,916 @@
+    /// `poll` may end up taking awhile, the work should be offloaded to a
+    /// thread pool (or something similar) to ensure that `poll` can return
+    /// quickly.
+    ///
+    /// # Panics
+    ///
+    /// Once a future has completed (returned `Ready` from `poll`), subsequent
+    /// calls to `poll` may panic, block forever, or otherwise cause bad behavior.
+    /// The `Future` trait itself provides no guarantees about the behavior of
+    /// `poll` after a future has completed.
+    fn poll(self: PinMut<Self>, cx: &mut task::Context) -> Poll<Self::Output>;
+}
+```
+
+Most of the explanation here follows what we've already said about the task
+system. The one twist is the use of `PinMut`, which makes it possible to keep data
+borrowed across separate calls to `poll` (i.e., "borrowing over yield
+points"). The mechanics of pinning are explained
+in [the RFC that introduced it](https://github.com/rust-lang/rfcs/pull/2349),
+and the interoperation with traditional futures is described next.
+
+## Relation to the futures crate
+
+In many respects this RFC simply imports a minimal slice of
+the [futures 0.2 API](http://aturon.github.io/2018/02/27/futures-0-2-RC/) into
+libcore, and indeed
+an [earlier iteration](https://github.com/rust-lang/rfcs/pull/2395) was
+explicitly just that. This incarnation, however, focuses squarely on the
+absolute minimal footprint needed to provide async/await support.
+
+It's thus out of scope for this RFC to say *precisely* what the futures 0.3 APIs
+will look like, but there are a few key aspects worth calling out. (Note that
+over time, many of these APIs will be stabilized into `std`.)
+
+### Error handling
+
+It's very common to work with `Result`-producing futures, so the crate will
+provide the following alias:
+
+```rust
+/// A convenience for futures that return `Result` values that includes
+/// a variety of adapters tailored to such futures.
+pub trait TryFuture {
+    /// The type of successful values yielded by this future
+    type Item;
+
+    /// The type of failures yielded by this future
+    type Error;
+
+    /// Poll this `TryFuture` as if it were a `Future`.
+    ///
+    /// This method is a stopgap for a compiler limitation that prevents us from
+    /// directly inheriting from the `Future` trait; in the future it won't be
+    /// needed.
+    fn try_poll(self: PinMut<Self>, cx: &mut task::Context) -> Poll<Result<Self::Item, Self::Error>>;
+}
+
+impl<F, T, E> TryFuture for F
+    where F: Future<Output = Result<T, E>>
+{
+    type Item = T;
+    type Error = E;
+
+    fn try_poll(self: PinMut<Self>, cx: &mut task::Context) -> Poll<F::Output> {
+        self.poll(cx)
+    }
+}
+```
+
+This alias makes it easy to require that a future return a `Result` (by bounding
+by `TryFuture`), and to obtain the success/error types (by using the `Item` and
+`Error` associated types).
+
+Similarly, `PollResult<T, E>` is a type alias for `Poll<Result<T, E>>`.
```

Comment:
This is probably pedantic, but 'similarly' seems wrong here - it is a completely different mechanism